### PR TITLE
Add OLMo-core based DPO training module

### PR DESCRIPTION
**Pull request by finbarrtimbers**
Adds an Olmo-core based DPO implementation in `dpo.py`. 

Runs:
- Single GPU DPO: [Beaker](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFE6VFWN0PAQYYAKR0Q05TX3?taskId=01KFE6VFWS3N57HKHN26YRF0WM&jobId=01KFE6VG0GBBZTHJDGJAXTYJ9E)
- Multi-node DPO: [Beaker](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFGSNPAC0G6J0HHF27H9YGJ4?)
- Also ran single GPU GRPO based off the result of running single GPU DPO:   [Beaker](https://beaker.org/ex/01KFHVH2DVX7YP7E32YAJCJSXH)

GPU_TESTS=[01KFH82YPAC7ECB4XXEHH10TS7](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFH82YPAC7ECB4XXEHH10TS7?taskId=01KFH82YPFK5359Z9KPNX4NNGF&jobId=01KFH82YT4N97CZNWHDAHP233Y)

### CODE REVIEW

**Comment by finbarrtimbers** [comment]
## Experiment Results

Ran single GPU DPO script ([Beaker](https://beaker.org/ex/01KFE6VFWN0PAQYYAKR0Q05TX3)) successfully.

Multi-node DPO experiments are experiencing Beaker cluster rendezvous timeout issues (infrastructure-related, not code issues). Will re-run when cluster stability improves.

File: open_instruct/dpo.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+from open_instruct import data_loader as data_loader_lib
+from open_instruct import dataset_transformation, dpo_utils, logger_utils, model_utils, utils
+from open_instruct.beaker_callback import BeakerCallbackV2
+from open_instruct.olmo_core_train_modules import DPOTrainModule
+from open_instruct.padding_free_collator import TensorDataCollatorWithFlatteningDPO
+
+logger = logger_utils.setup_logger(__name__)
+
+
+OLMO_MODEL_CONFIG_MAP: dict[str, str] = {
```
Comment (hamishivi) [inline open_instruct/dpo.py]: If this is going to be re-used in future (for sft/grpo), could we put it into some olmo-core utils file?

File: open_instruct/dpo.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    if args.gradient_checkpointing:
+        from olmo_core.nn.transformer.config import TransformerActivationCheckpointingMode
+
+        logger.info("Enabling activation checkpointing...")
+        model.apply_activation_checkpointing(TransformerActivationCheckpointingMode.full)
+
+    return model
+
+
+def _apply_hsdp(model, device: torch.device):
```
Comment (hamishivi) [inline open_instruct/dpo.py]: Can we add things like sequence parallel sharding in future? Are there other sharding strategies available in olmo-core we might want to use?

File: open_instruct/dpo_utils.py [hunk 3]
```diff
@@ -461,41 +466,55 @@ def compute_reference_cache_hash(args: ExperimentConfig, tc: TokenizerConfig) ->
...
     rejected_tensor = torch.full((full_dataset_size,), float("-inf"), dtype=torch.float32, device=device)
 
     with torch.no_grad():
-        for batch in tqdm(
-            dataloader, disable=not accelerator.is_local_main_process, desc="Caching reference logprobs"
-        ):
-            if use_lora:
-                with accelerator.unwrap_model(model).disable_adapter():
+        for batch in tqdm(dataloader, disable=not is_main_process, desc="Caching reference logprobs"):
+            if use_lora and disable_adapter_context is not None:
```
Comment (hamishivi) [inline open_instruct/dpo_utils.py:537]: is lora even supported in olmo-core? Does the use_lora flag do anything now?

File: scripts/train/debug/dpo/single_gpu.sh
Comment (hamishivi) [inline scripts/train/debug/dpo/single_gpu.sh:1]: Could we put these under `debug/dpo`? And could you add a non-mason script? I think its useful (a) for people wondering how to locally debug on interactives; (b) for people external to ai2 to see a script without mason.

File: open_instruct/dpo.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    )
+
+    forward_fn = dpo_utils.concatenated_forward_olmo if args.concatenated_forward else dpo_utils.separate_forward_olmo
+    if args.packing:
+        forward_fn = partial(dpo_utils.concatenated_forward_olmo, packing=True)
+    average_log_prob = args.loss_type.is_average_loss
+
+    logger.info("Caching reference logprobs (before HSDP)...")
+
+    def make_disable_adapter_context() -> contextlib.AbstractContextManager:
```
Comment (hamishivi) [inline open_instruct/dpo.py]: same comment as elsewhere: do we actually support lora for olmo-core, and should we have this conditional if so...?

File: open_instruct/dpo.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+        device=device,
+        cache_path=reference_cache_path,
+        is_main_process=is_main_process,
+        use_lora=args.use_lora,
+        disable_adapter_context=make_disable_adapter_context if args.use_lora else None,
+    )
+    logger.info("Reference logprobs cached.")
+    data_loader.reshuffle(epoch=0)
+
+    model = _apply_hsdp(model, device)
```
Comment (hamishivi) [inline open_instruct/dpo.py]: Does this mean the model isnt sharded for building the reference logprobs cache? What about models we can't fit onto 1 gpu even for inference?

File: open_instruct/dpo_utils.py (hunk 3 above)
Comment (finbarrtimbers) [reply open_instruct/dpo_utils.py:537]: No, this was AI slop 😅. Sorry. Has been removed!

File: open_instruct/dpo.py (hunk 4 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py]: No, this was AI slop 😅. Sorry. Has been removed!

File: open_instruct/dpo.py (hunk 5 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py]: ahhh nice catch. Fixing this!

File: scripts/train/debug/dpo/single_gpu.sh
Comment (finbarrtimbers) [reply scripts/train/debug/dpo/single_gpu.sh:1]: Good idea, doing both now!

File: open_instruct/dpo.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py]: yes! I added tensor, context, and pipeline paralellism.

**Comment by finbarrtimbers** [comment]
## DPO Experiment Results

Ran single GPU DPO script ([Beaker](https://beaker.org/ex/01KFGQEF2JF18M9GXHFXAFQF15)) and multi-node DPO ([Beaker](https://beaker.org/ex/01KFGRAS4DHK3BYERXX0DY1VRJ)) scripts.

### Results:
- **Single GPU DPO**: ✅ Passed (exit code 0)
- **Multi-node DPO**: ✅ Training completed successfully with checkpoints saved. Had a non-critical exit code 1 due to a multi-node barrier failure in post-training cleanup (a known issue with multi-node jobs where one node finishes before the other).

### Changes in this commit:
- Move generic OLMo code to `olmo_core_utils.py`
- Add support for tensor, context, and pipeline parallelism (`tensor_parallel_degree`, `context_parallel_degree`, `pipeline_parallel_degree`)
- Add LoRA error check (not supported with OLMo-core)
- Fix critical bug: apply HSDP before computing reference logprobs cache
- Fix race condition in reference logprobs cache directory creation
- Reorganize DPO scripts to `scripts/train/debug/dpo/`
- Add local.sh for testing without Beaker

**Comment by finbarrtimbers** [comment]
> Mostly some comments. Some other things:
> 
> 1. could we add a single-gpu script that runs locally? I tried `uv run torchrun     --standalone     --nproc_per_node=1     open_instruct/dpo.py     --model_name_or_path allenai/OLMo-2-0425-1B     --tokenizer_name allenai/OLMo-2-0425-1B     --use_flash_attn false     --max_seq_length 1024     --per_device_train_batch_size 1     --gradient_accumulation_steps 4     --learning_rate 5e-07     --lr_scheduler_type linear     --warmup_ratio 0.1     --weight_decay 0.0     --num_epochs 3     --output_dir output/dpo_olmo_core_debug/     --logging_steps 1     --mixer_list allenai/tulu-3-wildchat-reused-on-policy-8b 100     --chat_template_name olmo     --seed 123     --try_launch_beaker_eval_jobs false` but it (a) errored initially after building the cache and then (b) hung on further training. Just make a beaker image with `beaker://ai2/cuda12.8-dev-ubuntu22.04-notorch` and try run with uv to recreate.
> 2. It looks like the multi-node job ran okay but exited with an error? Is that fixable?

Fixed the multi-node job and adding a local script!

File: open_instruct/dpo.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py]: Yes, done!

File: open_instruct/dpo.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py]: yes, done!

File: scripts/train/debug/dpo/local.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+torchrun --nproc_per_node=1 open_instruct/dpo.py \
```
Comment (hamishivi) [inline scripts/train/debug/dpo/local.sh]: i think should be uv run torchrun?

File: open_instruct/dpo.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        num_replicas=None,
+        shard_degree=None,
+        param_dtype=DType.bfloat16,
+        reduce_dtype=DType.float32,
+        wrapping_strategy=TransformerDataParallelWrappingStrategy.blocks,
+    )
+
+    tp_config = tensor_parallel_degree if tensor_parallel_degree > 1 else None
+    cp_config = context_parallel_degree if context_parallel_degree > 1 else None
+    pp_config = pipeline_parallel_degree if pipeline_parallel_degree > 1 else None
```
Comment (hamishivi) [inline open_instruct/dpo.py:155]: 🤩🤩🤩🤩

File: scripts/train/debug/dpo/multi_node.sh
Comment (hamishivi) [inline scripts/train/debug/dpo/multi_node.sh:1]: Could we run something like this for a while and confirm the DPO is working well? Or e.g. rerun instruct olmo 3 dpo training and compare the results?

File: open_instruct/dpo.py [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        )
+    if args.with_tracking:
+        trainer_callbacks["wandb"] = callbacks.WandBCallback(
+            name=args.run_name or args.exp_name,
+            project=args.wandb_project,
+            entity=args.wandb_entity,
+            config=json_config,
+        )
+    checkpointing_steps = int(args.checkpointing_steps)
+    trainer_callbacks["checkpointer"] = CheckpointerCallback(save_interval=checkpointing_steps, save_async=False)
```
Comment (hamishivi) [inline open_instruct/dpo.py:237]: How does resuming jobs work in this setup? Is resumption seamless wrt beaker preemption?

File: open_instruct/dpo.py [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+        wandb_url = None
+        if args.with_tracking:
+            wandb_tracker = trainer_callbacks.get("wandb")
+            if wandb_tracker is not None and hasattr(wandb_tracker, "run") and wandb_tracker.run is not None:
+                wandb_url = wandb_tracker.run.get_url()
+        if args.hf_repo_revision is not None:
+            eval_path = args.output_dir
+            if beaker_config is not None and beaker_config.beaker_dataset_ids:
+                eval_path = beaker_config.beaker_dataset_ids[-1]
+            utils.launch_ai2_evals_on_weka(
```
Comment (hamishivi) [inline open_instruct/dpo.py:279]: Our evals assume the model is vllm-compatible, but I think now we save in the Olmo-core format? Shouldn't we convert back to HF for this + the hf upload?

File: open_instruct/olmo_core_utils.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    "allenai/OLMo-2-0325-32B": "olmo2_32B",
+    "allenai/Olmo-3-1025-7B": "olmo3_7B",
+    "allenai/OLMoE-1B-7B-0924": "olmoe_1B_7B",
+    "Qwen/Qwen3-0.6B": "qwen3_0_6B",
+    "Qwen/Qwen3-1.7B": "qwen3_1_7B",
+    "Qwen/Qwen3-4B": "qwen3_4B",
+    "Qwen/Qwen3-8B": "qwen3_8B",
+    "Qwen/Qwen3-14B": "qwen3_14B",
+    "Qwen/Qwen3-32B": "qwen3_32B",
+}
```
Comment (hamishivi) [inline open_instruct/olmo_core_utils.py:30]: A higher-level question: let's say I wanted to muck around with something that i need to go to olmo-core for (e.g., adding a new model). How easy is this without having to make and wait for official PRs? Is there an easy way to do it, and can we document it somewhere?

File: scripts/train/debug/dpo/local.sh (hunk 6 above)
Comment (finbarrtimbers) [reply scripts/train/debug/dpo/local.sh]: ah ty, fixed!

File: open_instruct/dpo.py (hunk 7 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py:155]: Let's just hope it works well!

File: scripts/train/debug/dpo/multi_node.sh
Comment (finbarrtimbers) [reply scripts/train/debug/dpo/multi_node.sh:1]: I propose that we merge this PR, and then rerun Olmo3 instruct DPO and compare the results.

File: open_instruct/dpo.py (hunk 8 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py:237]: yes!

File: open_instruct/olmo_core_utils.py (hunk 10 above)
Comment (finbarrtimbers) [reply open_instruct/olmo_core_utils.py:30]: If you need to add a new model, then we need to add code to `Olmo-core` (https://github.com/allenai/OLMo-core/pull/534, https://github.com/allenai/OLMo-core/pull/533). We don't need to merge the PR, as we can always point `open-instruct` to a branch of Olmo-core.

Where would you like it documented?

File: scripts/train/debug/dpo/multi_node.sh
Comment (hamishivi) [reply scripts/train/debug/dpo/multi_node.sh:1]: sure. Could we run + test a smaller model beforehand? I don't really mind what, we should have a small DPO job we trust the results on to some degree just to see it work.

File: open_instruct/olmo_core_utils.py (hunk 10 above)
Comment (hamishivi) [reply open_instruct/olmo_core_utils.py:30]: Good question... we need to do a docs update 😅 I think somewhere in the readme we should have something like a "developing" section where we can point to olmo-core for where much of the modelling code is, and then have a quick pointer on how to locally dev if you need to edit both at the same time.

Comment (finbarrtimbers) [reply open_instruct/olmo_core_utils.py:30]: I added a section to `CONTRIBUTING.md`. PTAL!

File: open_instruct/dpo.py (hunk 9 above)
Comment (finbarrtimbers) [reply open_instruct/dpo.py:279]: Done! I ran a GRPO experiment that runs off of the trained DPO model. Works fine.

File: scripts/train/debug/dpo/multi_node.sh
Comment (finbarrtimbers) [reply scripts/train/debug/dpo/multi_node.sh:1]: Sure. I added `7b_instruct_dpo_olmo_core.sh` and have an experiment running: https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFK2F7WEY5S46R8K8E492ZT5?taskId=01KFK2F7WK40TD9NND1RSRDTAJ&jobId=01KFK2F80RM0WJG3KTHF2Z3FKE

File: open_instruct/olmo_core_utils.py (hunk 10 above)
Comment (hamishivi) [reply open_instruct/olmo_core_utils.py:30]: Where did you add this? can't see.
